### PR TITLE
Fix #377 - Use spec() on DiscoveryInfo records

### LINKED ISSUE #377: Labeling schema should also be applied to task.DiscoveryInfo.Name
See:

https://github.com/mesosphere/mesos-dns/blob/e65fcabd5e5c0f7e21d65fd927914514e6d99a5f/records/generator.go#L380-L391

mesos-dns will apply `spec` on `task.Name`, but not on `task.DiscoveryInfo.Name`, this leads to inconsistent naming if a task specifies `DiscoveryInfo.Name`.

### PR DESCRIPTION
**Pull request by sargun**

### CODE REVIEW

**Comment by sargun** [comment]
@jdef

File: records/generator.go [hunk 1]
```diff
@@ -363,78 +363,100 @@ func (rg *RecordGenerator) listenerRecord(listener string, ns string) {
...
+		slaveIDTail(task.SlaveID),
+		task.IP(ipSources...),
+		task.SlaveIP,
+	}
 
-			rg.insertRR(arec+tail, ctx.taskIP, "A")
-			rg.insertRR(canonical+tail, ctx.taskIP, "A")
+	// use DiscoveryInfo name if defined instead of task name
+	if task.HasDiscoveryInfo() {
+		// LEGACY TODO: REMOVE
```
Comment (jdef) [inline records/generator.go]: this doesn't look right (this wasn't there in the old code)

Comment (jdef) [inline records/generator.go]: compare #382

File: records/generator_test.go [hunk 2]
```diff
@@ -196,6 +196,8 @@ func TestInsertState(t *testing.T) {
 		name string
 		want []string
 	}{
+		{rg.As, "big-dog.marathon.mesos.", []string{"10.3.0.1"}},
```
Comment (jdef) [inline records/generator_test.go]: where are the SRV's?

**Comment by jdef** [comment]
This looks broken to me. Some additional context in the description/header of this PR would be very nice.

File: records/generator_test.go (hunk 2 above)
Comment (sargun) [inline records/generator_test.go]: Added the SRV records to the test.

File: records/generator.go (hunk 1 above)
Comment (sargun) [inline records/generator.go]: Basically what I did break up the `tasksRecords` function into three functions -

`taskRecord` - Meant to operator on a single task. Before there was the usage of anonymous struct, ctx, for passing around informaiton about how a record should be built up, so I defined it as an independent type. Now, tasksRecord passed the task + framework to tasksRecord to taskRecord iif it was RUNNING, and we had an IP for the slave it was running on.

File: records/generator.go [hunk 3]
```diff
@@ -363,78 +363,100 @@ func (rg *RecordGenerator) listenerRecord(listener string, ns string) {
...
-		fname := labels.DomainFrag(f.Name, labels.Sep, spec)
 
-		// insert taks records
-		tail := "." + domain + "."
 		for _, task := range f.Tasks {
 			var ok bool
 			task.SlaveIP, ok = rg.SlaveIPs[task.SlaveID]
 
-			// skip not running or not discoverable tasks
-			if !ok || (task.State != "TASK_RUNNING") {
```
Comment (sargun) [inline records/generator.go]: Previously we would bail, and now we check for both.

File: records/generator.go [hunk 4]
```diff
@@ -363,78 +363,100 @@ func (rg *RecordGenerator) listenerRecord(listener string, ns string) {
...
+	slaveID,
+	taskIP,
+	slaveIP string
+}
 
-			// use DiscoveryInfo name if defined instead of task name
-			if task.HasDiscoveryInfo() {
-				ctx.taskName = task.DiscoveryInfo.Name
-			}
+func (rg *RecordGenerator) taskRecord(task state.Task, f state.Framework, domain string, spec labels.Func, ipSources []string) {
```
Comment (sargun) [inline records/generator.go]: Task Record just defines the context, and passes it to taskContextRecord.

File: records/generator.go [hunk 5]
```diff
@@ -363,78 +363,100 @@ func (rg *RecordGenerator) listenerRecord(listener string, ns string) {
...
-				slaveTarget := slaveHost + ":" + port
+}
+func (rg *RecordGenerator) taskContextRecord(ctx context, task state.Task, f state.Framework, domain string, spec labels.Func) {
+	fname := labels.DomainFrag(f.Name, labels.Sep, spec)
 
-				if !task.HasDiscoveryInfo() {
-					rg.insertRR(tcpName+tail, slaveTarget, "SRV")
-					rg.insertRR(udpName+tail, slaveTarget, "SRV")
-				}
+	tail := "." + domain + "."
```
Comment (sargun) [inline records/generator.go]: Github Diff formatting seems to have brokens with a combination of whitespace changes + reorganization of functions.

File: records/generator_test.go [hunk 6]
```diff
@@ -208,6 +210,14 @@ func TestInsertState(t *testing.T) {
 		{rg.As, "slave.mesos.", []string{"1.2.3.10", "1.2.3.11", "1.2.3.12"}},
 		{rg.As, "some-box.chronoswithaspaceandmixe.mesos.", []string{"1.2.3.11"}}, // ensure we translate the framework name as well
 		{rg.As, "marathon.mesos.", []string{"1.2.3.11"}},
+		{rg.SRVs, "_big-dog._tcp.marathon.mesos.", []string{
+			"big-dog-4dfjd-0.marathon.mesos.:80",
+			"big-dog-4dfjd-0.marathon.mesos.:443",
+		}},
+		{rg.SRVs, "_big-dog2._tcp.marathon.mesos.", []string{
```
Comment (jdef) [inline records/generator_test.go]: I think this is the (unintended?) behavior change. the "LEGACY" block you added causes this to be created. Prior to this PR the `big-dog2` task name is replaced by the `big-dog` discovery info name.

What's the justification for this change in behavior?

**Comment by sargun** [comment]
This was the solve the problem of both having the old behaviour when discoveryInfo was found as part of a Mesos task record (to put the record, denormalized) into DNS. What I've done is made to sure it adds both the normalized, and denormalized records. The statement in question is here:
```

			// use DiscoveryInfo name if defined instead of task name
			if task.HasDiscoveryInfo() {
				ctx.taskName = task.DiscoveryInfo.Name
			}
```
It should be:

```

			// use DiscoveryInfo name if defined instead of task name
			if task.HasDiscoveryInfo() {
				ctx.taskName = spec(task.DiscoveryInfo.Name)
			}
```

Unfortunately, this could break people's existing installations, so I split out the function into the first part which is just initialization, and loops. The second part, which is setting up the context, and the third part which is creating the records. If discovery info is detected, I do call the record generation component, before and after spec has been called on DiscoveryInfo.

File: records/generator.go (hunk 1 above)
Comment (jdef) [inline records/generator.go]: if you're going for backwards compat, then you're missing a line here in the LEGACY block, prior to invoking taskContextRecord: `ctx.taskName = task.DiscoveryInfo.Name`

File: records/generator_test.go (hunk 6 above)
Comment (sargun) [inline records/generator_test.go]: The idea is to continue to provide the existing behaviour + new behaviour:

```

		// LEGACY TODO: REMOVE
		rg.taskContextRecord(ctx, task, f, domain, spec)
		// LEGACY, TODO: REMOVE

		ctx.taskName = spec(task.DiscoveryInfo.Name)
		rg.taskContextRecord(ctx, task, f, domain, spec)
```

So, when the Marathon team adds a bunch of DiscoveryInfo entries, the old ones still exist.

Comment (jdef) [inline records/generator_test.go]: I think we're talking past each other. In my mind when there's discovery
info if you want to preserve the old behavior then you need ctx.taskName =
task.DiscoveryInfo.Name before taskContextRecord. for the new behavior, set
ctx.taskName=spec(task.DiscoveryInfo.Name) and call taskContextRecord again.

Comment (sargun) [inline records/generator_test.go]: Ah, okay, I think I misunderstood the original ask.

File: records/generator.go (hunk 1 above)
Comment (sargun) [inline records/generator.go]: Changed it to look more like described.

Comment (jdef) [inline records/generator.go]: great. can we set some deadline for when we're allowed to remove the legacy block?

**Comment by jdef** [comment]
can we incorporate the changes from the fake JSON in #382 and see how tests are impacted?

**Comment by sargun** [comment]
Completed.

**Comment by jdef** [comment]
lgtm.

when can we remove the legacy block?

**Comment by sargun** [comment]
Whenever someone looks at this ticket after June: https://github.com/mesosphere/mesos-dns/issues/400

**Pull request merged**